sympy/core/assumptions.py

```diff
@@ -543,7 +568,14 @@ def __init__(cls, *args, **kws):
         defs.update(local_defs)
 
         cls._explicit_class_assumptions = defs
-        cls.default_assumptions = StdFactKB(defs)
+        if 'Expr' not in [c.__name__ for c in cls.mro()]:
```

Comment:
```suggestion
        from sympy.core.expr import Expr
        if not isinstance(c, Expr):
```

Reply:
There should be a better mechanism for this rather than special casing Expr here.

Reply:
I think it would be better to use a separate metaclass for Expr e.g.:
```python
class BasicMeta(type):
    """Root metaclass for all Basic"""
    pass

class FactKB:
    """Class that handles logic inference for assumption systems"""
    pass

class EmptyKB(FactKB):
    """Logic inference system with no assumptions predicates"""
    pass

class ExprFactKB(FactKB):
    """Assumptions system for Expr with standard predicates"""
    pass

class BasicManagedProperties(BasicMeta):
    """Actual metaclass for Basic"""
    _FACTKB = EmptyKB

# The metaclass of a subclass must be a subclass of the metaclass of the
# superclass. This means that if Expr is a subclass of Basic then the metaclass
# for Expr must be a subclass of the metaclass for Basic.
class ExprManagedProperties(BasicManagedProperties):
    """Actual metaclass for Expr"""
    _FACTKB = ExprFactKB

class Basic(metaclass=BasicManagedProperties):
    """Root superclass for all symbolic things"""
    pass

class Expr(Basic, metaclass=ExprManagedProperties):
    """Root superclass for all expression-like things"""
    pass

```